### ISSUE TITLE
SPAMS dependency / is spams lasso the slowest part in the whole thing?

### ISSUE THREAD

**Issue by jgamper** [Peter554/StainTools#18]
We could replace spams lasso with a small NN based mapping to the concentrations. That should speed up the augmentor. What do you think @Peter554 ?

**Comment by Peter554** [comment]
Well you're more than welcome to try - won't know how effective this would be until then. If it works we could keep both methods and leave choice up to user?

**Comment by Peter554** [comment]
I would also be interested to swap out SPAMS for a more standard dependency like sklearn if possible. Presently I use SPAMS for two tasks:

- Lasso regression in `get_concentrations` for `FancyNormaliser`.
- Dictionary learning in Vahadane method. 

SPAMS has the nice functionality that you can constrain variables to be positive when dictionary learning (http://spams-devel.gforge.inria.fr/doc-python/html/doc_spams004.html#sec5) which is needed for Vahadane and I don't see this in sklearn (http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.DictionaryLearning.html)? For the lasso regression we also want a positivity constraint but I believe this is possible in both SPAMS (http://spams-devel.gforge.inria.fr/doc-python/html/doc_spams005.html#sec15) and sklearn (http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html).

Whether SPAMS  or sklearn I prefer to use the same for both tasks for consistency and keeping number of dependencies down - that's why I use SPAMS at the moment.

**Comment by jgamper** [comment]
label as discussion?

**Comment by Peter554** [comment]
SPAMS is now available via PyPI test site and will hopefully be on PyPI soon so closing for now

**Comment by andreped** [comment]
@jgamper did you find any solutions that sped up the augmentor? I've been testing out Macenko augmentation in my pipeline, but sadly its just too slow for me to properly utilize the GPU. Would be nice to have this augmentor as a tensorflow-layer or similar to run this more efficiently, but then on GPU. But perhaps I'll look into other faster alternatives. Currently, HSV is the way to go, but I feel like its just a little too näive for H&E-stained WSIs
